src/Core/Termination/CallGraph.idr

```diff
@@ -134,56 +137,107 @@ mutual
+                Core SizeChange
 
-  sizeCompareCon : Term vars -> Term vars -> Bool
-  sizeCompareConArgs : Term vars -> List (Term vars) -> Bool
-  sizeCompareApp : Term vars -> Term vars -> SizeChange
+  sizeCompareCon : {auto defs : Defs} -> Nat -> Term vars -> Term vars -> Core Bool
+  sizeCompareTyCon : {auto defs : Defs} -> Term vars -> Term vars -> Bool
+  sizeCompareConArgs : {auto defs : Defs} -> Nat -> Term vars -> List (Term vars) -> Core Bool
+  sizeCompareApp : {auto defs : Defs} -> Nat -> Term vars -> Term vars -> Core SizeChange
 
-  sizeCompare s (Erased _ (Dotted t)) = sizeCompare s t
-  sizeCompare _ (Erased _ _) = Unknown -- incomparable!
+  sizeCompare fuel s (Erased _ (Dotted t)) = sizeCompare fuel s t
+  sizeCompare fuel _ (Erased _ _) = pure Unknown -- incomparable!
   -- for an as pattern, it's smaller if it's smaller than either part
-  sizeCompare s (As _ _ p t)
-      = knownOr (sizeCompare s p) (sizeCompare s t)
-  sizeCompare (As _ _ p s) t
-      = knownOr (sizeCompare p t) (sizeCompare s t)
-  sizeCompare s t
-     = if sizeCompareCon s t
-          then Smaller
-          else knownOr (sizeCompareApp s t) (if sizeEq s t then Same else Unknown)
-
-  sizeCompareCon s t
+  sizeCompare fuel s (As _ _ p t)
+      = knownOr (sizeCompare fuel s p) (sizeCompare fuel s t)
+  sizeCompare fuel (As _ _ p s) t
+      = knownOr (sizeCompare fuel p t) (sizeCompare fuel s t)
+  -- if they're both metas, let sizeEq check if they're the same
+  sizeCompare fuel s@(Meta _ _ _ _) t@(Meta _ _ _ _) = pure (if sizeEq s t then Same else Unknown)
+  -- otherwise try to expand RHS meta
+  sizeCompare fuel s@(Meta n _ i args) t = do
+    Just gdef <- lookupCtxtExact (Resolved i) (gamma defs) | _ => pure Unknown
+    let (PMDef _ [] (STerm _ tm) _ _) = definition gdef | _ => pure Unknown
+    tm <- substMeta (embed tm) args zero []
+    sizeCompare fuel tm t
+    where
+      substMeta : {0 drop, vs : _} ->
+                  Term (drop ++ vs) -> List (Term vs) ->
+                  SizeOf drop -> SubstEnv drop vs ->
+                  Core (Term vs)
+      substMeta (Bind bfc n (Lam _ c e ty) sc) (a :: as) drop env
+          = substMeta sc as (suc drop) (a :: env)
+      substMeta (Bind bfc n (Let _ c val ty) sc) as drop env
+          = substMeta (subst val sc) as drop env
+      substMeta rhs [] drop env = pure (substs drop env rhs)
+      substMeta rhs _ _ _ = throw (InternalError ("Badly formed metavar solution \{show n}"))
+
+  sizeCompare fuel s t
+     = if sizeCompareTyCon s t then pure Same
+       else if !(sizeCompareCon fuel s t)
+          then pure Smaller
+          else knownOr (sizeCompareApp fuel s t) (pure $ if sizeEq s t then Same else Unknown)
+
+  -- consider two types the same size
+  sizeCompareTyCon s t =
+    let (f, args) = getFnArgs t in
+    let (g, args') = getFnArgs s in
+    case f of
+      Ref _ (TyCon _ _) cn => case g of
+        Ref _ (TyCon _ _) cn' => cn == cn'
+        _ => False
+      _ => False
```

Comment:
It is not clear to me whether it is sound to make applied type constructor of the same name size-equal regardless of their arguments. But I did not manage to come up with a counter-example (assuming a putative Idris with universe levels/universe polymorphism).

Do you have an intuition why this is correct?

Reply:
I think at the time I was considering them irrelevant.  I'm asking for things to be the same or smaller when we go under a constructor, and that was my way of ignoring the implicit type arguments.

I think we would have been ok here for now because we don't look under `TyCon` to determine something is smaller.  So if the type got bigger it could still never cause the function to be counted as total. 

I've changed it to check the arguments, expecting everything to be the `Same` or `Smaller`. It's still contributing `Same` at best. So, like the status quo, pulling something out from under a `TyCon` doesn't count as `Smaller`.
```idris
  sizeCompareTyCon fuel s t =
    let (f, args) = getFnArgs t in
    let (g, args') = getFnArgs s in
    case f of
      Ref _ (TyCon _ _) cn => case g of
        Ref _ (TyCon _ _) cn' => if cn == cn'
            then (Unknown /=) <$> sizeCompareProdConArgs fuel args' args
            else pure False
        _ => pure False
      _ => pure False
```